### DISCUSSION TITLE
:class directive not updating when state changed in x-init

### DISCUSSION THREAD

**Discussion by danjac** [alpinejs/alpine#1382] [7. Bugs (archived)]
I have a menu component which should change the class depending on the position in the viewport. This is calculated inside x-init. The class should update depending on the value of a "position" attribute.

The position value is calculated correctly, however the class as specified in the :class directive is not added to the element.

```
<div class="absolute left-0 bg-white dark:bg-gray-800 w-56 border shadow-xl rounded-lg z-20 mt-2"
     :class="{'top-0': position === 'dropdown', 'bottom-0': position === 'dropup'}"
     x-data="{show: true, position: 'dropdown'}"
     x-show="show"
     @click.away="show=false"
     @keydown.escape.window="show=false"
     x-init="() => {
          const rect = $el.getBoundingClientRect();
          const vh = document.documentElement.clientHeight;
          position = (vh - rect.top - rect.height) > 0 ? 'dropdown' : 'dropup';
          // console.log(position) 
    }">
 
```

**Comment by KevinBatdorf** [comment]
This should probably be a discussion. 

That code works as expected for me when I added it to codepen. Can you reproduce it in a demo where it's not working? 

I think the issue you're seeing is that the classes are applied first (expected behavior) which affects the output of `getBoundingClientRect`. So if you change your starting position in `x-data` you'll always get the same result back.

**Comment by danjac** [comment]
Likewise unable to replicate the problem in the codepen. The issue I'm seeing is that neither "top-0" or "bottom-0" are inserted, so even if get`BoundingClientRect()` is returning an unexpected result, the class list should be updated with one of those values depending on `position`. In fact if I remove the `x-init() `entirely the "top-0" is not inserted into the class list. 

I'll see if I can find more debugging info.

**Comment by danjac** [comment]
So looks like it's unrelated to Alpine: I'm swapping the outer HTML of the <div> (using htmx), and the class list is actually updated correctly before being replaced (not entirely sure why as the element replaced does not have an x-data attribute). In any case Alpine worked correctly once that issue was fixed.
